### PR TITLE
Add buy now feature.

### PR DESCRIPTION
**Pull request by orionstardust**
ref: https://github.com/aavegotchi/aavegotchi-gbm-baazaar/pull/21

### CODE REVIEW

File: contracts/facets/GBMFacet.sol [hunk 1]
```diff
@@ -181,6 +178,81 @@ contract GBMFacet is IGBM, IERC1155TokenReceiver, IERC721TokenReceiver, Modifier
...
+        if (a.highestBid > 0) {
+            IERC20(s.GHST).transfer(a.highestBidder, a.highestBid + a.dueIncentives);
+            //emit incentive event and bidRemoval event
+            emit Auction_IncentivePaid(_auctionID, a.highestBidder, a.dueIncentives);
+            emit Auction_BidRemoved(_auctionID, a.highestBidder, a.highestBid);
+        }
+
+        emit Auction_BoughtNow(_auctionID, msg.sender);
+
+        if (IERC165(ca).supportsInterface(0x2a55205a)) {
```
Comment (cinnabarhorse) [inline contracts/facets/GBMFacet.sol:229]: Please refactor this code so we're not duplicating code (it also appears in the "claim" function) @orionstardust

File: contracts/facets/GBMFacet.sol [hunk 2]
```diff
@@ -420,11 +493,34 @@ contract GBMFacet is IGBM, IERC1155TokenReceiver, IERC721TokenReceiver, Modifier
...
+            a.buyItNowPrice = _buyItNowPrice;
+        } else {
+            a.buyItNowPrice = 0;
+        }
+
+        emit Auction_BuyItNowUpdated(_auctionID, _buyItNowPrice);
+    }
+
+    //Recommended to be set to 70
+    function setBuyItNowInvalidationThreshold(uint256 _invalidationThreshold) external onlyOwner {
```
Comment (cinnabarhorse) [inline contracts/facets/GBMFacet.sol:520]: This function should be called atomically in the upgrade script @orionstardust

File: contracts/facets/GBMFacet.sol [hunk 3]
```diff
@@ -143,42 +140,56 @@ contract GBMFacet is IGBM, IERC1155TokenReceiver, IERC721TokenReceiver, Modifier
...
-        if (a.info.tokenKind == ERC1155) {
-            _sendTokens(ca, recipient, ERC1155, tid, tam);
+        uint256 ae1bnp = a.buyItNowPrice;
+        if (ae1bnp == 0) revert("NoBuyItNowPrice");
+        if (((ae1bnp * s.buyItNowInvalidationThreshold) / 100) <= a.highestBid) revert("HighestBidTooHighToBuyNow");
+
+        if (msg.sender == a.owner) revert("OwnerBuyNowNotAllowed");
+        if (a.info.startTime > block.timestamp) revert("AuctionNotStarted");
+        //verify existence
+        if (a.owner == address(0)) revert("NoAuction");
```
Comment (cinnabarhorse) [inline contracts/facets/GBMFacet.sol:168]: Duplicated code @orionstardust

File: contracts/facets/GBMFacet.sol [hunk 4]
```diff
@@ -247,25 +254,36 @@ contract GBMFacet is IGBM, IERC1155TokenReceiver, IERC721TokenReceiver, Modifier
...
         s.auctionNonce++;
+
+        //In order to start an auction with a minium starting price, you need to prepay the fees
+        if (_info.startingBid != 0) {
+            //Transfer the money of the seller to the GBM Diamond
+            uint256 prepaidFee = (_info.startingBid * 40) / 1000; //4% fee, hardcoded
+            IERC20(s.GHST).transferFrom(msg.sender, address(this), prepaidFee);
+
+            //Presettle the fee
+            uint256 _rem = _settleFees(_info.startingBid);
```
Comment (cinnabarhorse) [inline contracts/facets/GBMFacet.sol:265]: Why are we using settleFees and not setleFeesWithRoyalty here? @orionstardust

Comment (orionstardust) [reply contracts/facets/GBMFacet.sol:265]: @cinnabarhorse 
starting bid is a fee. this is why we use `_settleFees`

Comment (cinnabarhorse) [reply contracts/facets/GBMFacet.sol:252]: Got it

**Pull request merged**